src/libutil/file-system.hh

```diff
@@ -141,6 +141,19 @@ inline bool symlink_exists(const std::filesystem::path & path) {
     return std::filesystem::exists(std::filesystem::symlink_status(path));
 }
 
+/**
+ * Canonicalize a path except for the last component.
+ *
+ * This is useful for getting the canonical location of a symlink.
+ *
+ * Consider the case where `foo/l` is a symlink. `canonical("foo/l")` will
+ * resolve the symlink `l` to its target.
+ * `parent_canonical("foo/l")` will not resolve the symlink `l` to its target,
+ * but does ensure that the returned parent path's parent equals its `canonical`,
+ * and can therefore be retrieved without traversing any symlinks.
+ */
+std::filesystem::path parent_canonical(const std::filesystem::path & path);
```

Comment:
Let's not add more stuff to this `nix::fs` namespace, it has no reason to exist. And to be consistent with our naming convention:
```suggestion
std::filesystem::path canonicalParent(const std::filesystem::path & path);
```

Reply:
Done. It's `makeParentCanonical` now, so that it doesn't suggest returning the parent.